File: ice0208/graphs-dfs/547.py
class Solution:
    def findCircleNum(self, isConnected: list[list[int]]) -> int:
        provinces = 0

        visit = [False] * len(isConnected)
        for i in range(len(isConnected)):
            if visit[i]:
                continue
            provinces += 1
            stack = [i]

            while stack:
                cur = stack.pop()
                visit[cur] = True
                for idx, con in enumerate(isConnected[cur]):
                    if con == 1 and visit[idx] == False:
                        stack.append(idx)

        return provinces


if __name__ == "__main__":
    testcase_list = [
        [[1, 1, 0], [1, 1, 0], [0, 0, 1]],
        [[1, 0, 0], [0, 1, 0], [0, 0, 1]],
    ]

    for testcase in testcase_list:
        print(Solution().findCircleNum(testcase))

File: moonpoison/graphs_dfs/547.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected)
    {
        int v[201]={0};
        stack<int> s;
        int cnt=0;
        for(int q=0;q<isConnected.size();++q)
        {
            if(!v[q])
            {
                for(int i=0;i<isConnected[q].size();++i)
                    if(isConnected[q][i]&&i!=q) s.push(i);
                v[q]++;
                while(!s.empty())
                {
                    int a=s.top();
                    s.pop();
                    if(!v[a])
                    {
                        v[a]++;
                        for(int i=0;i<isConnected[a].size();++i)
                            if(isConnected[a][i]&&i!=a) s.push(i);
                    }
                } 
                cnt++;
            }
        }
        return cnt;
    }
};